### ISSUE TITLE
New nodes should be send past X hours of history before current tuples

### ISSUE THREAD

**Issue by rixed** [rixed/ramen#552]
Alternatives:

- Workers could attach themselves (via an RPC to the supervisor?) so that they could first request the history then the realtime;
- Workers could select from a proximate tuple provider that would adapt the history and then either forward the real output or reconnect the child (via supervisor?) Could be a single proxy per supervisor or one per parent;
- producer could retrieve its own history itself, retrieve required fields from it and feed out-ringbuffer ((only) advantage seems to be that it has the native-code functions to unserialize/serialize its output).
- À la Aurora: have all messages go through a MQ in all cases, thus acting as a huge proxy all the time;
- If a max history length is set per producer rather than/in addition to per consumer (and that max length should fit in 1 RB), we could have a proxy running alongside the producer and always keep that history in RAM thus cutting down on IOs ;
- Or actually the producer itself could do it, with the help of a more sophisticated output function, that both adds the current output to the history and then flush that history down the throat of the children from their respective last seqnums;
- Twice as many processes: producer always write all output fields into a single proxy (which input ringbuf size is the max history length) and which handles all the output distribution.
- Completely different direction: Replay events from source! But then we have to fork new versions of the workers for we do not want to mix past aggregate groups with new ones (at least for those nodes that do perform grouping). Overall looks like more work but the advantages are:
  - Pro: can choose the history length on a case by case basis;
  - Pro: no need for each node to keep X hours of history in huge ringbufs;
  - Cons: no guarantee that we would recompute the same tuples (esp. in face of wallclock depending functions or merges/joins)

Depends on #556

**Comment by rixed** [comment]
Not needed now that timeseries extraction always make use of replay.
